lib/util/npm-util.js

```diff
@@ -0,0 +1,89 @@
+/**
+ * @fileoverview Utility for executing npm commands.
+ * @author Ian VanSchooten
+ * @copyright 2016 Ilya Volodin. All rights reserved.
+ * See LICENSE file in root directory for full license.
+ */
+
+"use strict";
+
+//------------------------------------------------------------------------------
+// Requirements
+//------------------------------------------------------------------------------
+
+var execSync = require("child_process").execSync;
+
+//------------------------------------------------------------------------------
+// Helpers
+//------------------------------------------------------------------------------
+
+/**
+ * Add a key to an object with a value of false
+ * @param   {Object} obj  An Object
+ * @param   {string} key  A string to use as a key
+ * @returns {Object}      The same object as provided, with the key added, having a value of false.
+ */
+function addFalseKey(obj, key) {
+    obj[key] = false;
+    return obj;
+}
+
+//------------------------------------------------------------------------------
+// Private
+//------------------------------------------------------------------------------
+
+/**
+ * Install node modules synchronously and save to devDependencies in package.json
+ * @param   {string|string[]} packages Node module or modules to install
+ * @returns {void}
+ */
+function installSyncSaveDev(packages) {
+    if (Array.isArray(packages)) {
+        packages = packages.join(" ");
+    }
+    execSync("npm i --save-dev " + packages, {stdio: "inherit"});
+}
+
+/**
+ * Check whether a node module or set of node modules is installed within a project
+ * Assumes cwd is within the project.
+ *
+ * @param   {string|string[]} packages Node module or modules to check.
+ * @returns {Object}                   An object whose keys are the module names
+ *                                     and values are booleans indicating installation.
+ */
+function checkInstalled(packages) {
+    var pkgString,
+        pkgStatus,
+        result;
+
+    if (typeof packages === "string") {
+        pkgString = packages;
+        packages = [packages];
+    } else if (Array.isArray(packages)) {
+        pkgString = packages.join(" ");
+    }
+
+    pkgStatus = packages.reduce(addFalseKey, {});
+    try {
+        result = execSync("npm ls " + pkgString + " --json --depth=0", { stdio: ["ignore", "pipe", "ignore"], encoding: "utf8" });
+    } catch (err) {
+        // error is thrown if no matches were found
+        result = JSON.stringify({ dependencies: {} });
+    }
+
+    return packages.reduce(function(status, pkgName) {
+        status[pkgName] = Object.keys(JSON.parse(result).dependencies).indexOf(pkgName) > -1;
```

Comment:
We should extract the object keys and json parse outside the reduce to improve perf.